sky/resources.py

```diff
@@ -262,6 +262,9 @@ def is_launchable(self) -> bool:
 
     def need_cleanup_after_preemption(self) -> bool:
         """Returns whether a spot resource needs cleanup after preeemption."""
+        if self.cloud is None:
+            logger.info('self.cloud is None, assuming no cleanup is needed.')
```

Comment:
How about we enforce the caller to be a launchable resource?
```suggestion
        assert self.is_launchable(), self
```

Reply:
Yes sure, the reason of just printing logs is because I'm not sure whether we should fail the entire recovery process if we can't determine `need_cleanup_after_preemption`. adding assertion may fail the spot job. If we think this is okay then let's do assertion.

Reply:
I think the developer should make sure that the resources passed to this function are launchable. Otherwise, I don't think that makes much sense to return `False` for non-launchable resources.

Reply:
Yes I agree, I thought about returning `None` when it's not launchable. so developer should expect `None` to be returned if they pass a resource without cloud. tho this will further complicate `controller.py`. We can do assertion.